### ISSUE TITLE
Slow reload on Windows 10 using PyCharm

### ISSUE THREAD

**Issue by Kludex** [Kludex/uvicorn#2000]
### Discussed in https://github.com/encode/uvicorn/discussions/1977

<div type='discussions-op-text'>

<sup>Originally posted by **adhihargo** May 14, 2023</sup>
Uvicorn used to work fine in my dev laptop running Windows 10, but now every time there's a modified change, it stuck in this line:

`WARNING:  WatchFiles detected changes in 'main.py'. Reloading...`

... for nearly 5 minutes before actually reloading the files (printing `Shutting down` in the next line, etc.). During that time, the API server part works normally but serving the older version before file modification..

Downgraded to v0.21.1, and reloading works instantly again.</div>

> [!IMPORTANT]
> - We're using [Polar.sh](https://polar.sh/encode) so you can upvote and help fund this issue.
> - We receive the funding once the issue is completed & confirmed by you.
> - Thank you in advance for helping prioritize & fund our backlog.

<a href="https://polar.sh/encode/uvicorn/issues/2000">
<picture>
  <source media="(prefers-color-scheme: dark)" srcset="https://polar.sh/api/github/encode/uvicorn/issues/2000/pledge.svg?darkmode=1">
  <img alt="Fund with Polar" src="https://polar.sh/api/github/encode/uvicorn/issues/2000/pledge.svg">
</picture>
</a>

**Comment by 3to1null** [comment]
I think issue lies somewhere here: https://github.com/encode/uvicorn/blob/7d4eab20ac0cf0fb07b3afaa9ca35a2ff09fb499/uvicorn/supervisors/basereload.py#L87-L99

With the changes introduced in #1584, it will hang for multiple minutes on the process.join() call. If I remove the changes, so the function looks like:
```
    def restart(self) -> None:
        self.process.terminate()
        self.process.join()
        self.process = get_subprocess(
            config=self.config, target=self.target, sockets=self.sockets
        )
        self.process.start()
```
The reload is pretty much instant, like it was on earlier versions.

**Comment by 3to1null** [comment]
I can only reproduce this issue when the program is run using Pycharm. If I run it directly (either via `uvicorn.run("app:app", reload=True)` in Python, or with the `uvicorn app:app --reload` from the terminal), this issue is **not** present.

While it would be very nice to have this working again, I am not sure if this is something that should be fixed in Uvicorn (if this is even possible).

Edit: By enabling 'emulate terminal in output console' in Pycharm's edit run configuration dialog, the reload works as expected. However, this also seems to disable ANSI escape sequences, so the output logs look quite bad.
![image](https://github.com/encode/uvicorn/assets/16557274/842d3995-b2a9-4d9a-b328-0e1c7b1bff41)

**Comment by Kludex** [comment]
Got it.

Please open an issue on PyCharm's issue tracker. If they tell me it's a Uvicorn bug, and explain why, we can think about the next step. 👍

**Comment by 3to1null** [comment]
Done: https://youtrack.jetbrains.com/issue/PY-60962/os.killpid-signal.CRTLCEVENT-ignored-when-running-python-program-from-Pycharm.

I've added a very small program that has the same issue and does not use Uvicorn. I do not think this is a Uvicorn bug, however if Jetbrains does not want to fix this, in my opinion it might still be worth it to investigate if we can support this in Uvicorn by not using `os.kill(self.process.pid, signal.CTRL_C_EVENT)`. Having a working reloader and being able to use Pycharm is probably the expected behaviour for a lot of developers.

btw, thank you for the quick responses :)

**Comment by rcoor** [comment]
Hello, I am getting a similar error on 0.22.0 but it happens both when I run with uvicorn.run or in the terminal, with --reload.

**Comment by altq33** [comment]
The same problem right now on FastAPI and default pycharm run settings with --reload

**Comment by Kludex** [comment]
Is there someone having this issue **without** pycharm?

**Comment by altq33** [comment]
> Is there someone having this issue **without** pycharm?

No, this problem occurs when starting the server through pycharm tools, if you run the server with the same command through the terminal, --reload works instantly

**Comment by Kludex** [comment]
Thanks @altq33 :)

I'm locking this issue since people keep spamming here.

This is a **pycharm** issue: [youtrack.jetbrains.com/issue/PY-60962/os.killpid-signal.CRTLCEVENT-ignored-when-running-python-program-from-Pycharm](https://youtrack.jetbrains.com/issue/PY-60962/os.killpid-signal.CRTLCEVENT-ignored-when-running-python-program-from-Pycharm).

If you have an issue with Windows (not using Pycharm), please create a new issue.

(I'll not close this issue so people can be aware)

**Comment by Kludex** [comment]
I've unlocked the issue, so people can concentrate the complains here instead of creating multiple issues/discussions.

**Comment by leonardomarcao** [comment]
@Kludex @altq33 @3to1null any resolution? I have the same problem... :'(
I having troubleshoot using the Uvicorn an FasAPI in debugger mode.  

Some information about my environment:

+ Poetry venv
+ Python 3.12
+ last versions of FastAPI, uvicorn etc.

**Comment by Kludex** [comment]
Not trying to be rude but... Use VSCode, or complain to JetBrains.

The issues on PyCharm's side were created and ignore, as far as I'm aware.

**Comment by 3to1null** [comment]
@leonardomarcao The issue in Pycharm still persists and is tracked here: https://youtrack.jetbrains.com/issue/PY-60962/os.killpid-signal.CRTLCEVENT-ignored-when-running-python-program-from-Pycharm

**Comment by soulee-dev** [comment]
Still struggling same issue.
Should be changed...

**Comment by Kludex** [comment]
Locking this issue again. This is a PyCharm issue, not Uvicorn.

Here: https://youtrack.jetbrains.com/issue/PY-60962/os.killpid-signal.CRTLCEVENT-ignored-when-running-python-program-from-Pycharm

**Comment by jamesh1999** [comment]
I've been doing some research on this as it's causing me some annoyance. The issue is neither limited to `watchfiles` nor PyCharm. I get the same issue with the StatReloader using:
- VSCode's built in terminal
- The new Windows terminal app (running Powershell or CMD)
- And PyCharm

It seems to be due to Windows not passing on the Ctrl+C event until the console is updated in some way. This leaves you in the weird situation where, it will be stuck at "Reloading..." until you go and click the terminal window, bringing it to the foreground.

I've found quite an effective (but hacky) method to mitigate this: if you simply `print("")` after sending the `CTRL_C_EVENT` in `basereload.py`, it also sufficiently updates the console to trigger the event to get handled.

I would prefer not to revert back to `self.process.terminate()` as it is imperative for me that the process exits cleanly.

@Kludex in case this is of use

**Comment by Rajrup-TransEV** [comment]
def restart(self) -> None:
        self.process.terminate()
        self.process.join()
        self.process = get_subprocess(
            config=self.config, target=self.target, sockets=self.sockets
        )
        self.process.start() this solution is working thanks a lot

**Comment by kitsupixel** [comment]
> I've been doing some research on this as it's causing me some annoyance. The issue is neither limited to `watchfiles` nor PyCharm. I get the same issue with the StatReloader using:
> 
> * VSCode's built in terminal
> * The new Windows terminal app (running Powershell or CMD)
> * And PyCharm
> 
> It seems to be due to Windows not passing on the Ctrl+C event until the console is updated in some way. This leaves you in the weird situation where, it will be stuck at "Reloading..." until you go and click the terminal window, bringing it to the foreground.
> 
> I've found quite an effective (but hacky) method to mitigate this: if you simply `print("")` after sending the `CTRL_C_EVENT` in `basereload.py`, it also sufficiently updates the console to trigger the event to get handled.
> 
> I would prefer not to revert back to `self.process.terminate()` as it is imperative for me that the process exits cleanly.
> 
> [@Kludex](https://github.com/Kludex) in case this is of use

As someone only using VSCode on Windows, this was also a problem for me, but the print("") solved it.
Thanks very much!

**Comment by Kludex** [comment]
- Closed on #2604
